theano/link/basic.py

```diff
@@ -262,14 +268,28 @@ class LocalLinker(Linker):
 
     """
 
-    def make_thunk(self, input_storage=None, output_storage=None, storage_map=None):
+    def make_thunk(
+        self,
+        input_storage: Optional[Any] = None,
+        output_storage: None = None,
+        storage_map: Optional[
+            Union[
+                Dict[
+                    Union["TensorVariable", "TensorConstant"],
+                    Union[List[None], List[ndarray]],
+                ],
+                Dict["TensorConstant", List[ndarray]],
+            ]
+        ] = None,
```

Comment:
This is a compound type that is entirely worth defining as its own type.  I added such definitions [here](https://github.com/brandonwillard/Theano-PyMC/blob/bd5f63ccb76860e707dc2077804a16eaaf75aacc/theano/graph/op.py#L52), but there's probably a better way to do this (e.g. using [`NewType`](https://docs.python.org/3/library/typing.html#newtype)).

Reply:
Following the example, the variable type are defined after import as:

```python
StorageMapType = Union[
    Dict[Union["TensorVariable", "TensorConstant"], Union[List[None], List[ndarray]]],
    Dict["TensorConstant", List[ndarray]],
]
OutputStorageType = List[Optional[List[Any]]]
```

Reply:
Yeah, that `StorageMapType` looks more accurate.

Reply:
Actually, I think `StorageMapType` should actually use `Variable` and/or `Constant` instead of the more narrow `TensorVariable` and `TensorConstant` types.

Reply:
It's also possible that `ndarray` isn't the only value type.  There are a few special types for non-`ndarray` types (e.g. `None` and `slice` types).

Reply:
I have added `ThunkType` as:

```python
ThunkType = Tuple[Callable[[], NoReturn], List["Container"], List["Container"]]
```

I am a bit unsure concerning `StorageMapType`, I have redefined as:
```python
StorageMapType = Union[
    Dict[Union[Variable, Constant], Union[List[None], List[ndarray], List[slice]]],
    Dict[Constant, Union[List[None], List[ndarray], List[slice]]],
]
```

Reply:
You don't need `Union[Variable, Constant]`, because `Constant` is a `Variable` subclass.

Something like `StorageMapType = Dict[Variable, List[Optional[Union[ndarray, slice]]]]` should suffice for now.

Reply:
Done